### ISSUE TITLE
Add dark mode 

### ISSUE THREAD

**Issue by acorreya** [8VIM/8VIM#41]
Please add a night mode theme

**Comment by acorreya** [comment]
This is how it looks with force dark mode is enabled. 
![Screenshot_20201016-034258_Fennec](https://user-images.githubusercontent.com/3225091/96191648-26607780-0f62-11eb-8464-f829ee1c5e19.png)

**Comment by flide** [comment]
too much demand for dark mode!! will have to understand the damn thing i guess!!

**Comment by LunarEclipse363** [comment]
Hi, I found this project recently and loved the idea! I've noticed some issues when trying it out, but since it's open-source and I have some programming experience (mostly non-Java though) I decided to try messing around with the codebase to see if I can help with it. 

After spending some time on it today I managed to get most of the app themed dark, making all the colors refer to the theme so it should be possible to easily add new themes or switch them. I didn't really figure out how to theme the actual keyboard yet, but I haven't really familiarized myself with all the Java code so I'll look there next.

If you want to take a look at the code, I will push what you can see on the attached screenshot to [the `theming` branch on my fork](https://github.com/Mckol/8VIM/tree/theming) soon. As mentioned I don't really have much android dev experience so if there's anything wrong please tell me 🙂

Here's a screenshot of how it looks right now:
![Screenshot_20201104-205108_8Vim](https://user-images.githubusercontent.com/24840010/98169348-5f01ca00-1eec-11eb-969e-2074580319fa.png)

**Comment by acorreya** [comment]
This is wonderful! I think I will look in to the code too, but it will be a
while till i get here, given my side hustles.

Thank you flide and Mckol!

**Comment by mynery** [comment]
I would give it a try, but ... what about the keyboard itself? That's what i will see most of the time anyways.

**Comment by LunarEclipse363** [comment]
> I would give it a try, but ... what about the keyboard itself? That's what i will see most of the time anyways.

I'm trying to figure that out rn, for some reason when I apply a theme to the xboard layout, it crashes when I open the keyboard. 

I also tried to find a way to programmatically get a color from the current theme to set the colors (something I'd do in XML like `?android:colorBackground`) but none seem to really work.

I'm gonna look into implementing an actual theme switching mechanism, maybe that would make it easier to get the colors from the current theme.

I will send more information on what I've tried and the crash later today when I'm working on this again.

**Comment by LunarEclipse363** [comment]
Here's a snippet from `8vim/src/main/java/inc/flide/vim8/views/mainKeyboard/XboardView.java`, it has comments with the details on what I've tried in regards to getting colors and how did it work out, also a bit about the crash:
The `bg_color` and `fg_color` declarations are duplicated just to keep syntax highlighting in the snippet, I of course only tried one method at a time.
```java
public class XBoardView extends View {
    [...]
    @SuppressLint("RestrictedApi") // For some reason MaterialColors.getColor triggers this lint.
    private void initialize(Context context) {
        // The below line seems to be where it crashes if a theme is applied in
        // `8vim/src/main/res/layout/x_keyboard_layout.xml`
        // The error I get will be attached below this snippet.
        actionListener = new MainKeyboardActionListener((MainInputMethodService) context, this);
        setHapticFeedbackEnabled(true);

        // Those seem to work with `@SuppressLint("RestrictedApi")` on the method as shown above, 
        // but as said previously actually applying a theme makes the app crash before this gets applied.
        // If I don't apply a theme and use this, the keyboard is a uniform gray box, but it still works.
        @ColorInt int bg_color = MaterialColors.getColor(context, android.R.attr.colorBackground, Color.BLACK);
        @ColorInt int fg_color = MaterialColors.getColor(context, android.R.attr.colorForeground, Color.MAGENTA);

        // Those give me a "Expected resource of type color" error in the second argument
        @ColorInt int bg_color = ContextCompat.getColor(context, android.R.attr.colorBackground);
        @ColorInt int fg_color = ContextCompat.getColor(context, android.R.attr.colorBackground);

        backgroundPaint.setColor(bg_color);
        foregroundPaint.setColor(fg_color);
        foregroundPaint.setAntiAlias(true);
        foregroundPaint.setStrokeJoin(Paint.Join.ROUND);
        foregroundPaint.setTextSize(Constants.TEXT_SIZE);
        Typeface font = Typeface.createFromAsset(getContext().getAssets(),
                "SF-UI-Display-Regular.otf");
        foregroundPaint.setTypeface(font);
    }
    [...]
}
```
And the error from debugger console when applying a theme:
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: inc.flide.vi8, PID: 3970
    android.view.InflateException: Binary XML file line #2: Binary XML file line #2: Error inflating class inc.flide.vim8.views.mainKeyboard.MainKeyboardView
    Caused by: android.view.InflateException: Binary XML file line #2: Error inflating class inc.flide.vim8.views.mainKeyboard.MainKeyboardView
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:647)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:790)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:730)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
        at inc.flide.vim8.MainInputMethodService.onCreateInputView(MainInputMethodService.java:52)
        at android.inputmethodservice.InputMethodService.updateInputViewShown(InputMethodService.java:1275)
        at android.inputmethodservice.InputMethodService.showWindowInner(InputMethodService.java:1669)
        at android.inputmethodservice.InputMethodService.showWindow(InputMethodService.java:1637)
        at android.inputmethodservice.InputMethodService$InputMethodImpl.showSoftInput(InputMethodService.java:480)
        at android.inputmethodservice.IInputMethodWrapper.executeMessage(IInputMethodWrapper.java:196)
        at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:37)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: android.view.InflateException: Binary XML file line #22: Binary XML file line #2: Error inflating class inc.flide.vim8.views.mainKeyboard.XboardView
     Caused by: android.view.InflateException: Binary XML file line #2: Error inflating class inc.flide.vim8.views.mainKeyboard.XboardView
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:647)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:790)
        at android.view.LayoutInflater.parseInclude(LayoutInflater.java:965)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:859)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:824)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:515)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
        at inc.flide.vim8.views.mainKeyboard.MainKeyboardView.initialize(MainKeyboardView.java:56)
        at inc.flide.vim8.views.mainKeyboard.MainKeyboardView.<init>(MainKeyboardView.java:35)
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:647)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:790)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:730)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
        at inc.flide.vim8.MainInputMethodService.onCreateInputView(MainInputMethodService.java:52)
        at android.inputmethodservice.InputMethodService.updateInputViewShown(InputMethodService.java:1275)
E/AndroidRuntime:     at android.inputmethodservice.InputMethodService.showWindowInner(InputMethodService.java:1669)
        at android.inputmethodservice.InputMethodService.showWindow(InputMethodService.java:1637)
        at android.inputmethodservice.InputMethodService$InputMethodImpl.showSoftInput(InputMethodService.java:480)
        at android.inputmethodservice.IInputMethodWrapper.executeMessage(IInputMethodWrapper.java:196)
        at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:37)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.ClassCastException: android.view.ContextThemeWrapper cannot be cast to inc.flide.vim8.MainInputMethodService
        at inc.flide.vim8.views.mainKeyboard.XboardView.initialize(XboardView.java:78)
        at inc.flide.vim8.views.mainKeyboard.XboardView.<init>(XboardView.java:64)
        	... 33 more
```

**Comment by flide** [comment]
Can you push your changes to a branch and raise a pr? i will try and have a
look.

**Comment by LunarEclipse363** [comment]
Sure I'll raise a WIP PR. Also I'm starting to think I'm doing something wrong here, I'll look at how other projects do theming and see if I can learn something that way.

**Comment by MichalLebeda** [comment]
@flide can be closed, theming supported in #172
